Dev: Add BIT user attributes to MS User

mtreacy002 opened the pull request:
### Description

Adding BIT user attributes to MS User as the first step for full BIT-MS integration

Fixes #1073 

### Type of Change:
- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)


### How Has This Been Tested?
The local_data.db has successfully added BIT User specific attributes after running migration script.

<img width="1438" alt="Screen Shot 2021-04-18 at 1 49 33 pm" src="https://user-images.githubusercontent.com/29667122/115133610-031b9f80-a04d-11eb-8219-5c5661e8ab95.png">

### Checklist:
- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**

- [ ] My changes generate no new warnings 
- [ ] New and existing unit tests pass locally with my changes

mtreacy002 committed:
Dev: Add BIT user attributes to MS User table

remove BIT attrributes from being returned in _repr_(self)

remove unnecessary brackets

Fix description

add bit branch to github action

fix black linting error

Remove alembic auto-generated comments

epicadk commented on app/utils/bitschema_utils.py:
```diff
@@ -0,0 +1,599 @@
+
+
+@unique
+class Timezone(Enum):
```
Is this required? There should be something inbuilt to do the same thing. I know there is in java.

mtreacy002 replied:
I'm open to refactoring this Timezone enum to a more efficient way if someone can pointed out an inbuilt package that is available out there. We can try to refactor the Timezone in the BIT repo first instead of here to see if it works coz BIT schema already the tests written for the endpoints that are using this Timezone enum.

epicadk replied:
https://www.iana.org/time-zones

epicadk replied:
Also [this](https://stackoverflow.com/questions/44836984/is-there-a-reason-for-java-time-zoneid-not-including-an-enum-of-zoneids#:~:text=3%20Answers&text=I%20believe%20it%27s%20because%20the,change%20regardless%20of%20Java%20version.&text=If%20the%20timezone%20data%20is,leaving%20the%20API%20%22incomplete%22.) because we use java for the android repo

epicadk replied:
Instead of this enum we could do [this](https://docs.python.org/3/library/zoneinfo.html) it's an inbuilt package. But implementing this means that the tz package would have to be the same for all the application.

mtreacy002 replied:
@epicadk , unfortunately, this `zoneinfo` module only available for Python >= 3.9. 
<img width="1030" alt="Screen Shot 2021-04-22 at 10 38 41 am" src="https://user-images.githubusercontent.com/29667122/115639247-7415cc80-a357-11eb-9b06-1c8583fc07a0.png">


MS backend is still using Python 3.6 for gh-action, so the gh action build/test will fail unless we change this.
<img width="609" alt="Screen Shot 2021-04-22 at 10 41 35 am" src="https://user-images.githubusercontent.com/29667122/115639407-d8389080-a357-11eb-91a8-8ff692acd309.png">

Also, this enum is currently used in BIT frontend in [timezones.js](https://github.com/anitab-org/bridge-in-tech-web/blob/develop/src/timezones.js) as a constant so that we can provide the dropdown options to user to prevent them inserting timezone that doesn't fall within the accepted value of backend timezone enum (which is according to the [pytz list](https://gist.github.com/heyalexej/8bf688fd67d7199be4a1682b3eec7568)). Do you have any suggestions how we do this if we refactor backend to `zoneinfo`? What should we use as constant at the frontend?

epicadk replied:
We'll run a matrix build. I'll create an issue.  Mentorship system runs fine on python version 3.9.

isabelcosta replied:
@mtreacy002 can you now use python 3.9 and use zoneinfo?

epicadk commented on app/database/models/user.py:
```diff
@@ -106,11 +118,20 @@ def json(self):
     def __repr__(self):
-        """Returns the user's name and username. """
-        return f"User name {self.name} . Username is {self.username} ."
+        """Returns the user's name, username, whether or not they are an organization representative and their timezone. """
```
Requires approval from @isabelcosta

mtreacy002 replied:
Yep..., I actually think it won't be too much of a problem if we don't include the `is_organization_rep` or `Timezone` in here since they can be abstracted later once we grab the user data using `GET /user` api call 🤔. I will just remove them for now 😁.

mtreacy002 commented:
My approached in this PR was from an angle of a new contributor who has succesfully completed the local setup.
The followings are the steps done on this PR:

1. updated local `develop` branch by pulling the latest commit from upstream.

2. created a new pr branch

3. run `python run.py` comand => this created the local_data.db. Notice that the `alembic version` folder din't exist in the db.
<img width="1431" alt="Screen Shot 2021-04-18 at 12 20 31 pm" src="https://user-images.githubusercontent.com/29667122/115133840-838ed000-a04e-11eb-8052-606545ee758b.png">

4. Registered a new user and confirm email (this is now what most contributors would have on their db where they have data but no `alembic version` folder as contributors have never run migration script before.

5. run initial migration using command "flask db migrate -m "Initial migration" => this added `alembic version` folder into the local_data.db.
<img width="1437" alt="Screen Shot 2021-04-18 at 12 33 02 pm" src="https://user-images.githubusercontent.com/29667122/115134279-209f3800-a052-11eb-9dc5-88b95376f3cb.png">

6. Added the `render_as_batch` condition in `run.py` so that running `flask-db-upgrade` won't return this error
<img width="552" alt="Screen Shot 2021-04-18 at 2 10 59 pm" src="https://user-images.githubusercontent.com/29667122/115133996-e7fe5f00-a04f-11eb-961c-2f01f7e8442f.png">

<img width="428" alt="Screen Shot 2021-04-18 at 2 08 12 pm" src="https://user-images.githubusercontent.com/29667122/115133958-850cc800-a04f-11eb-8360-1532e72313cb.png">

7. Made the schema changes in UserModel by adding BIT user specific attribute. Then run another migration command `flask db migrate -m "<message>"` where `<message>` is a clear message what the change was about. This was used as the alembic version title. This round running the script actually created the new alembic script version as it should.
<img width="327" alt="Screen Shot 2021-04-18 at 3 06 17 pm" src="https://user-images.githubusercontent.com/29667122/115134902-a40f5800-a057-11eb-9c38-f105fa28b384.png">
 
8. Fixed migration script so running `flask db upgrade` won't return this error
<img width="540" alt="Screen Shot 2021-04-18 at 3 14 16 pm" src="https://user-images.githubusercontent.com/29667122/115135048-c0f85b00-a058-11eb-85be-9e0bf28bec9e.png">

The culprit was shown below
<img width="738" alt="Screen Shot 2021-04-18 at 12 51 42 pm" src="https://user-images.githubusercontent.com/29667122/115134404-192c5e80-a053-11eb-914c-bd5a0ebfde0d.png">

And here's the fix
<img width="1436" alt="Screen Shot 2021-04-18 at 1 12 10 pm" src="https://user-images.githubusercontent.com/29667122/115135334-24838800-a05b-11eb-8d2f-5fe38f8514ad.png">

9. Refreshed the local_data.db on the sqlite db viewer tool and saw the new attributes were added (screenshot given in pr descripction above)

epicadk commented:
A tip for @isabelcosta and other reviewers you can the changes @mtreacy002 has made [here](https://github.com/anitab-org/mentorship-backend/pull/1074/commits/c6f2d5ef2685165d5e9e672e5188438e4202adc3)

mtreacy002 commented:
@epicadk, there were soooo many other commits there included in this PR 😱. Must be because the `bit` branch hasn't been synced to the `develop` branch from the admin/maintainers side. Is it possible for you to do this? I think once `bit` branch is updated, the irrelevant commits will disappear from the pr, right?

epicadk commented:
I think they will yes. Also can you please add BIT to github actions?

mtreacy002 commented:
> I think they will yes. Also can you please add BIT to github actions?

Done

mtreacy002 commented:
@epicadk, should I also add `bit` to gh-action docs.yml, or wait till we have docs related issue to add bit to the docs workflow?

mtreacy002 commented:
> # [Codecov](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org) Report
> > ❗ No coverage uploaded for pull request base (`bit@479f1c7`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#section-missing-base-commit).
> > The diff coverage is `99.84%`.
> 
> [![Impacted file tree graph](https://camo.githubusercontent.com/85d4ad8f36aa76cf01d7031b360bdfae6da82ba2ee1b147ae8903c4739cfd310/68747470733a2f2f636f6465636f762e696f2f67682f616e697461622d6f72672f6d656e746f72736869702d6261636b656e642f70756c6c2f313037342f6772617068732f747265652e7376673f77696474683d363530266865696768743d313530267372633d707226746f6b656e3d50425570524839337a712675746d5f6d656469756d3d726566657272616c2675746d5f736f757263653d6769746875622675746d5f636f6e74656e743d636f6d6d656e742675746d5f63616d706169676e3d70722b636f6d6d656e74732675746d5f7465726d3d616e697461622d6f7267)](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org)
> 
> ```diff
> @@          Coverage Diff           @@
> ##             bit    #1074   +/-   ##
> ======================================
>   Coverage       ?   94.39%           
> ======================================
>   Files          ?       39           
>   Lines          ?     2623           
>   Branches       ?        0           
> ======================================
>   Hits           ?     2476           
>   Misses         ?      147           
>   Partials       ?        0           
> ```
> 
> [Impacted Files](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org)	Coverage Δ	
> [app/api/models/mentorship_relation.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9tb2RlbHMvbWVudG9yc2hpcF9yZWxhdGlvbi5weQ==)	`100.00% <ø> (ø)`	
> [app/api/models/task.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9tb2RlbHMvdGFzay5weQ==)	`100.00% <ø> (ø)`	
> [app/utils/bitschema_utils.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL3V0aWxzL2JpdHNjaGVtYV91dGlscy5weQ==)	`99.83% <99.83%> (ø)`	
> [app/api/dao/mentorship_relation.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9kYW8vbWVudG9yc2hpcF9yZWxhdGlvbi5weQ==)	`96.11% <100.00%> (ø)`	
> [app/api/dao/task.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9kYW8vdGFzay5weQ==)	`98.48% <100.00%> (ø)`	
> [app/api/dao/task_comment.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9kYW8vdGFza19jb21tZW50LnB5)	`100.00% <100.00%> (ø)`	
> [app/api/dao/user.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9kYW8vdXNlci5weQ==)	`85.77% <100.00%> (ø)`	
> [app/api/resources/admin.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9yZXNvdXJjZXMvYWRtaW4ucHk=)	`88.13% <100.00%> (ø)`	
> [app/api/resources/task.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9yZXNvdXJjZXMvdGFzay5weQ==)	`100.00% <100.00%> (ø)`	
> [app/api/resources/user.py](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org#diff-YXBwL2FwaS9yZXNvdXJjZXMvdXNlci5weQ==)	`89.75% <100.00%> (ø)`	
> ... and [6 more](https://codecov.io/gh/anitab-org/mentorship-backend/pull/1074/diff?src=pr&el=tree-more&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=anitab-org)

@epicadk , can you suggest what went wrong here? 🤔
Is it just bcoz this is the first report so nothing to compare? That should be fine then, right?

epicadk commented:
> @epicadk, should I also add `bit` to gh-action docs.yml, or wait till we have docs related issue to add bit to the docs workflow?

Yup I think you can wait for this.

epicadk commented:
> @epicadk , can you suggest what went wrong here? 🤔
> Is it just bcoz this is the first report so nothing to compare? That should be fine then, right?

Yup that's right. It's fine.

mtreacy002 commented:
@isabelcosta , I have resolved the conflict, can you please review again 😉

isabelcosta commented:
auch @mtreacy002 can you squash all these commits into one? I see an issue here, and I bet it's because of the way that I did the sync 🙈 

![image](https://user-images.githubusercontent.com/11148726/115145485-a9fe3c80-a049-11eb-8f6c-edbcc2aa908b.png)

epicadk commented:
> auch @mtreacy002 can you squash all these commits into one? I see an issue here, and I bet it's because of the way that I did the sync 🙈
> 
> ![image](https://user-images.githubusercontent.com/11148726/115145485-a9fe3c80-a049-11eb-8f6c-edbcc2aa908b.png)

Nope rebasing won't be possible. It would still show the error.

mtreacy002 commented:
@isabelcosta , do you still face the conflict on merging? Do you want me to try squash merge all the commits from my end?

epicadk commented:
> @isabelcosta , do you still face the conflict on merging? Do you want me to try squash merge all the commits from my end?

You could try but I'm guessing it would still show merge conflicts after squashing. What you could do is. Rebase them into a single commit and then merge the upstream branch.

mtreacy002 commented:
Done, @isabelcosta and @epicadk , I've squashed and force-pushed the commits. Can you please check again if this still causing conflicts at your end when merging?

epicadk commented:
> Done, @isabelcosta and @epicadk , I've squashed and force-pushed the commits. Can you please check again if this still causing conflicts at your end when merging?

Yeps this might work.

isabelcosta commented on run.py:
```diff
@@ -14,7 +14,7 @@ def create_app(config_filename: str) -> Flask:
     db.init_app(app)
 
-    migrate = Migrate(app, db)
+    migrate = Migrate(app, db, render_as_batch=True)
```
what does this do? 🤓

isabelcosta commented on migrations/versions/1272e4b54a32_add_user_bitschema_attribute.py:
```diff
@@ -0,0 +1,638 @@
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
These comments from alembic can be removed right?

mtreacy002 replied:
@isabelcosta , done 😉

epicadk commented:
> 6. Added the `render_as_batch` condition in `run.py` so that running `flask-db-upgrade` won't return this error

@isabelcosta

mtreacy002 commented:
@vj-codes , I remember you asked last time whether we need to add `Black` to `requirements.txt` and my answered to that was that it's not needed. However, now that I encountered this issue with local `Black` version passing files but `Black` version in gh-action failing it, it seems fair to put `Black` to the `requirements.txt so that we can make sure all contributors running the same version locally that we'll match to the latest stable version on gh-action. What do you think of this? cc @epicadk

epicadk commented:
> @vj-codes , I remember you asked last time whether we need to add `Black` to `requirements.txt` and my answered to that was that it's not needed. However, now that I encountered this issue with local `Black` version passing files but `Black` version in gh-action failing it, it seems fair to put `Black` to the `requirements.txt so that we can make sure all contributors running the same version locally that we'll match to the latest stable version on gh-action. What do you think of this? cc @epicadk

Yup seems like a nice idea. But maybe add it to `dev-requirements.txt`? Also sorry for the delay I have some other commitments so I can't really attend the ms open session. I'll try to attend the upcoming one so we can discuss and possibly merge this. I actually think we can merge this and handle the timezone case later.

epicadk commented:
@isabelcosta no other changes from my side, can this be merged?

isabelcosta commented:
@mtreacy002 can you try to solve the merge conflicts please? Me, @vj-codes and @epicadk did some merges and branch recreation, and now that we added the matrix python version thing to GH Actions, there are conflicts.

mtreacy002 commented:
I've solved the conflict @isabelcosta 😉

isabelcosta reviewed (approved):
Will approve and merge once it has 2 approvals.

Points raised:
- We discussed that, timezones will be refactored later, to come from `zoneinfo` module.
- Instructions on how to run migrations should go to the README.

epicadk reviewed (approved):
👍🏼
